### ISSUE TITLE
Dose this library support lazy evaluation?

### ISSUE THREAD

**Issue by hyunsooda** [ramda/ramda#2356]
Hello, I'm university student in Korea.
I have a question to your lib.
As indicated above, I was wondering if your lib support lazy evaluation.
I have just written the example code, however It seems not working properly.

<img width="619" alt="2017-10-23 2 43 31" src="https://user-images.githubusercontent.com/12508269/31864656-fde6d39c-b79b-11e7-81c2-9c0f0afbde0c.png">

If your lib support, I think that first code have to work. but count was applyied fully.
Have a nice day.

**Comment by CrossEye** [comment]
First off, the library does not support lazy evaluation.  For the most general concept of lazy evaluation, the language won't support it.  For instance, in a lazy language, doing the equivalent of this operation:

```js
    [5 + 10, 21 + 21]
```

would not instantly evaluate `5 + 10` and `21 + 21`, since they might never be needed.  Javascript the language is eager, not lazy, and it cannot do that.  So Ramda, built on top of it, cannot either.

But there are libraries in Javascript that do some lazy list processing, [Lazy.js][la] and [lz.js][lz] among them.  These simply take advantage of such facts as

```js
compose(map(f), map(g)) <==> map(compose(f, g))
```

to make for more efficient iteration of lists.

Ramda has something similar in its [transducers][tr] [support][in], but it's not one of the core features of the library; you have to opt into it.

As to why your code doesn't stop at the right time: It does.  Your reporting mechanism is off.  You are updating the global variable `count` in both the `lazyEval` function and the `extract` one.  So it's being updated multiple times per iteration.  The whole thing stops when you hit 36 or so.

----------

One request, if you post other questions here (or likely in any GitHub project), don't post an *image* of your code.  Post the code itself.  (For details about how to make it readable, see https://guides.github.com/features/mastering-markdown/#examples, and choose the "code" tab)  Or if it's really large, post it to some code-sharing site and post a link here.  We recommend our [REPL][rr] for Ramda code, but there are many others for general Javsacript code, such as [JSBin][bi], [JSFiddle][fi], [RunKit][ru], [CodePen][co], and many others.  And if you don't need to run it, GitHub offers [gists][gi].

  
  [bi]: https://jsbin.com/
  [co]: https://codepen.io/
  [fi]: https://jsfiddle.net/
  [gi]: https://gist.github.com/
  [in]: http://ramdajs.com/docs/#into
  [la]: http://danieltao.com/lazy.js/
  [lz]: https://github.com/goatslacker/lz
  [ru]: https://runkit.com/
  [rr]: http://ramdajs.com/repl/
  [tr]: http://ramdajs.com/docs/#transduce

**Comment by hyunsooda** [comment]
I'm so sorry. I was not sure that culture exist in git. 
Lazy.js is good lib, but I would like to study with your lib.

Actually, Above code which be posted myself work as lazy evaluation but, the Lazyfunc is not common in order to another situation because that function was written by hardcoding.
So, I have tried to revise that code but, problems were generated.

first. 
R.take(3) << this function is curried so, parameter that pass into function is array type.
Therefore, Both parameter of getThird and R.take(3) were not same so, I'm not sure how give parameter into function.

second. 
R.take(3) is curried so, new function is created but, the way is not exist to see whether 3 or another.
therefore if parameter of one is passed like getThird function, not work properly.

I'm not certain that my english skills is so poor that whether you understand or not.

**Comment by kedashoe** [comment]
Hi @hyunsooda . Can you maybe post some code in the REPL and share a link? Here is a start:

https://goo.gl/FcvLFX

You can edit that and then click the "Share" link to get a URL to your code. Something else that always helps is letting everyone know your input and what you expect the output to be. For example

```js
// input
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

// exptected output
[0, 3, 6, 9]
```

These sorts of questions are great for our chat room on gitter, there are always lots of people there happy to help https://gitter.im/ramda/ramda

**Comment by CrossEye** [comment]
> I'm not certain that my english skills is so poor that whether you understand or not.

It's clear enough that you're not a native English speaker, but your questions are mostly understandable.

First, here's a version of your function that works in a lazy manner:

```js
const fn2 = into([], compose(getThird, getFourth, take(3)));
fn2(items); //=> [0, 12, 24]
```

[`into`][in] is part of Ramda's **Transducers** support.  You can find out more about them in #950 and #1269.  This works because `getThird` and `getFourth` are basically calls to `filter`, which already offers transducers support.

> R.take(3) << this function is curried so, parameter that pass into function is array type.
Therefore, Both parameter of getThird and R.take(3) were not same so, I'm not sure how give parameter into function.

I'm not sure if I understand this correctly.  I was under the impression that you wanted to do this:

    items                 [0, 1, 2, 3, ..., 100]
     `-> getThird         [0, 3, 6, 9, 12, ..., 99]
         `-> getFourth    [0, 12, 24, 35, 48, ..., 96]
             `-> take(4)  [0, 12, 24]

So the output of `getThird` should be of the type as the input to `getFourth` and the output of that should be of the same type as the input to `take(4)`.  Also `items` should be the type of the input to `getThird`.  But in fact, each of these types is an array of numbers, so things should just work.

I posted a version with your original code and the `into` one from above to https://goo.gl/vZQBVg.

That should demonstrate that the number of calls the second time is quite a bit smaller when using the transduced version.  This is all Ramda offers for lazy processing: some of its functions serve as transducers and we offer the mechanism to use our transducers or your custom ones.

> R.take(3) is curried so, new function is created but, the way is not exist to see whether 3 or another.
> therefore if parameter of one is passed like getThird function, not work properly.

I think you're asking about passing the `3` of `take(3)` as a parameter.  If so, then I think it would be difficult to do that in a point-free manner.

But it's very easy to do like this:

```js
const fn3 = (n, items) => into([], compose(getThird, getFourth, take(n)), items);
```

And you could wrap that in `curry` if you wanted to partially apply the first parameter.

By the same token, you can simplify your `extract` function when you're done verifying things with the `count` variable:

```js
let extract = (num) => R.filter(elem => elem % num === 0);
let getThird = extract(3), getFourth = extract(4);
```



  [in]: http://ramdajs.com/docs/#into

**Comment by hyunsooda** [comment]
Thanks for responding kindly.

I went to the link which provided by yourself and into function give good impression to me. 
By any chance, do you know mechanism of into that lazy evaluation worked.
I tried to watch the code in your api doc but, I did not read since they wrapped.

I still have not solved the original problem. this is implementation using your lib.
If functions which evaluated just predicate, I can come up with follow idea.

(  if call like this ->  lazyEval(getThird,getFourth,getFifth(3))  )
getThird(item[0]) -> getFourth(item[0]) -> getFifth(item[0])
getThird(item[1]) -> getFourth(item[1]) -> getFifth(item[1])
getThird(item[2]) -> getFourth(item[2]) -> getFifth(item[2])
......

However, R.take is not predicate and just function.
So, I think that function like take has to has predicate in order to working commonly.

If take function added,

(  if call like this ->  lazyEval(getThird,getFourth,take(3))  )
getThird(item[0]) -> getFourth(item[0]) -> take(item[0])
getThird(item[1]) -> getFourth(item[1]) -> take(item[1])
getThird(item[2]) -> getFourth(item[2]) -> take(item[2])
......
So, this is properly not working.


I was wondering if I have a defective knowledge on lazyEvaluation implementation.
I'm so sorry, it seems bother you.

**Comment by buzzdecafe** [comment]
> I posted a version with your original code and the into one from above to https://goo.gl/vZQBVg.

Great example, @CrossEye !

**Comment by CrossEye** [comment]
I think we are having some communication difficulties.

I'm not quite sure what you want.  Ramda has some built-in lazy processing capabilities, by using `transduce` and `into`.  This doesn't cover every function, but does cover a number of them. These ones are documented with the phrase, "Acts as a transducer if a transformer is given in list position."  Using these adds some complexity: `compose` works in a surprisingly reversed order, you need to add `into` and an initial value or `transduce`and several parameters.  But it does offer a consistent way to change a sequence of transformations of lists by operations to a single transformation on a list using a sequence of operations.  And that's a huge win.

This isn't true lazy processing.  That is something else entirely.  But for certain goals, it's a good substitute.  

Using this, `take(3)` is one of those steps.  It's a transformation of a list (or I guess a stream) of data into a fixed-length list.  But it's not a predicate.  If you want some transformation that uses predicates (say for filters) or other list transformations, you'll probably end up reproducing a lot of the infrastructure of transducers.  (If you want to just use predicates, you might consider [`allPass`][ap].)

But your initial code really did do some lazy processing.  It only called `extract` the number of times required to find the final output.  The trouble is that you had to encode the equivalent of `take(3)` into it.  Transducers would make it easier to build such things out of prebuilt functions.  But they simply encode that complex logic internally.  They don't eliminate it.

In other words, you can build your function out of things like `getThird`, `getFourth`, and `take` fairly straightforwardly with `into`.  Or you can write your own wrapper; but it won't be simple.

Does that help?

  [ap]: http://ramdajs.com/docs/#allPass

**Comment by hyunsooda** [comment]
Your all comments very useful and could be improved to by me.

I have just written the lazy evalulation once again.
I would like to write my idea about lazy evalulation.

Initially, I came up with a structure for lazy evalulation
<img width="903" alt="2017-10-27 2 51 23" src="https://user-images.githubusercontent.com/12508269/32068677-d0919e94-bac1-11e7-94a3-1fe4df2975fc.png">

After finalizing test code, compared mine to yours(into method) 
I noticed that mine was not completely fitted with take function.

I was under the impression that you designed like similar approching.
If take function is passed by into, It pass predicates smoothly before meet the 'take'
So, Item which be passed is stored with a array and execute when filled fully.
However, my lazy evaluation is different with yours and this was mistake.

If take function is passed by mine, according to the number that curried by take function, have to wait until fully filled and then execute rest.
So, the function made by myself think that location of take function is important since velocity can change.

Actually, in terms of lazy evalulation, if all parameter wrer simple predicates, It is not make concern.
However if, the function like 'take'  is added? according to the function, It have to can make logic and this is very hard with me.



This is code. actually you had better view the code more than listen my speaking in English.
http://jsbin.com/dorevir/edit?html,js,console


I am about to speak a number of word but, I don't come up idea suddenly. 
I appreciate your respond and I want your patience because I guess that my speaking is not courteous  because I'm not sure nuance or expression in English.

Thank you

**Comment by CrossEye** [comment]
I don't have time to look at your code right now; I will try to do so soon.  But I did want to respond to this:

> I appreciate your respond and I want your patience because I guess that my speaking is not courteous because I'm not sure nuance or expression in English.

You are clearly not quite fluent in English.  But I believe that you've usually been able to make yourself understood.  It's been a pleasure communicating with you.

**Comment by Izhaki** [comment]
@CrossEye 

> First off, the library does not support lazy evaluation.

I find it to be somewhat modest (in a bad way).

As you have shown, this will not iterate a list of 100k items:

```javascript
const fn2 = into([], compose(getThird, getFourth, take(3)));
```

In fact, my understanding of transducers is largely thanks to this [excellent Medium post that demonstrates lazy evaluation using ramda's transducers](https://medium.com/@roman01la/understanding-transducers-in-javascript-3500d3bd9624).

I have been using lazy.js in the past, and whilst Ramda doesn't support all of its features, it does supports the most important use cases. Many of lazy.js feature can be implemented using Ramda API.

So my view is that Ramda *does* support lazy evaluation.

**Comment by hyunsooda** [comment]
like your sentence, I think Javascript do not  completely accept functional programming's parts. I felt this when I was making the example code.
As alternative, I am considering study Clojure in the future.
your code of 4lines are also excellent description.
Thank you.

**Comment by CrossEye** [comment]
Javascript has many features which are not functional.  But you can use it as a mostly functional language.  Laziness is a feature of some functional languages, but a language can certainly be functional without being lazy (see LISP.)  Clojure is an interesting language, closer to the LISP end of the FP spectrum.  Haskell is on the opposite end.  Both are worth studying.

**Comment by hyunsooda** [comment]
I will enjoy developing my app using ur lib, lastly, thank you and have a nice day!

**Comment by buzzdecafe** [comment]
thanks for the stimulating discussion, @hyunsooda . You may also want to join the discussion in the ramda gitter room: https://gitter.im/ramda/ramda
